### DISCUSSION TITLE
Add licensing details to JavaScripts according to GNU LibreJS

### DISCUSSION THREAD

**Discussion by bhavin192** [plausible/analytics#719] [Anything Else]
Thank you for an awesome tool for website analytics!
About GNU LibreJS
> LibreJS is a free add-on for GNU IceCat and other Mozilla-based browsers. It blocks nonfree nontrivial JavaScript while allowing JavaScript that is free and/or trivial.

This basically requires the JS files to have a type of comments which specify the license, original source (if the file is minified), link etc. As far as I understood all the JS files we have are under AGPL license (and tracking one is in MIT), so it will be great to mark those files with the metadata. 

[Setting Your JavaScript Free](https://www.gnu.org/software/librejs/free-your-javascript.html)

Especially the tracking script can be easy one to start with. I'm don't know JavaScript much, but if required I can spend time and see what can be done to get this done (at least for the tracking script).

### For the `/settings` page
![Plausible_LibreJS](https://user-images.githubusercontent.com/5154532/107883209-9af8f200-6f13-11eb-945a-73b85621a121.png)

### For the tracking script
![tracking_Plausible](https://user-images.githubusercontent.com/5154532/107883554-af3dee80-6f15-11eb-88d3-df43a3d5b9ec.png)

**Comment by ukutaht** [comment]
You're correct in that the tracker is MIT licensed to avoid any virality issues on external pages.The rest of the product is licensed under AGPL.

Adding a comment will increase the size of the script (we are counting every single byte). Is there any other way to provide a license hint to LibreJS?

**Reply by bhavin192** [reply]
> Adding a comment will increase the size of the script (we are counting every single byte). Is there any other way to provide a license hint to LibreJS?

I agree. I was looking at the LibreJS documentation again, here is what I found.

There is a way to have a HTML page with table of JS files being used by a site / tool (let's call it weblabels page). Every page which loads the scripts should add a reference to the weblabels page. Example: https://weblabels.fsf.org/jitsi.member.fsf.org/CURRENT/ is being referenced by each page on https://jitsi.member.fsf.org/ using this:
```
<a href="https://weblabels.fsf.org/jitsi.member.fsf.org/CURRENT/" rel="jslicense">JavaScript license information</a>
```
If my website is loading `plausible.js`, then my site should have the weblabels page, or the JS should have comments; which we are not considering. The only option for pages having tracking code is to have a weblabels page, with license name, original source etc. (Edit: ~~I will confirm if this works by either trying myself or asking someone~~ it works that way).

Does this warrant a documentation page? Or should I just add some more details on what needs to be done to get the LibreJS allow `plausible.js`? That way people can implement it on their site.

OTOH, for other scripts used by Plausible admin interface either weblabels or comment method might be useful I think.

**Reply by ukutaht** [reply]
Thanks @bhavin192.

It's a bit unfortunate that these are the only options. I wish there was a way to add license hints as HTTP headers. That would probably be the most convenient for all.

> Does this warrant a documentation page? Or should I just add some more details on what needs to be done to get the LibreJS allow plausible.js? That way people can implement it on their site.

I have no idea how much interest there is for this. First time I've heard of LibreJS to be honest. I appreciate you putting the time in this but I'd need to see more people asking about LibreJS before taking the responsibility of maintaining official documentation on this. Until then, I would encourage writing a blog post or even just dropping the solution in this discussion so it's discoverable through search engines etc.

**Reply by bhavin192** [reply]
> It's a bit unfortunate that these are the only options. I wish there was a way to add license hints as HTTP headers. That would probably be the most convenient for all.

I agree. 

What you said about a documentation page for this, is totally valid. Once I'm done with my trials, I will post the solution for tracking script here.

**Comment by bhavin192** [comment]
@ukutaht what do you think about something like this for the admin interface of Plausible? https://invidious.snopyta.org/licenses
I don't know Elixir, but I can definitely try to implement a page like Invidious for Plausible. I can even take responsibility of maintaining that code :smile: This will involve creating a new page and making sure all the other pages have a reference to this page.

**Reply by ukutaht** [reply]
I think that would be great!
